fsgrid.hpp

```diff
@@ -614,28 +610,147 @@ template <typename T, int stencil> class FsGrid : public FsGridTools{
+         swap(first.neighbour_index, second.neighbour_index);
+         swap(first.ntasksPerDim, second.ntasksPerDim);
+         swap(first.taskPosition, second.taskPosition);
+         swap(first.periodic, second.periodic);
+         swap(first.globalSize, second.globalSize);
+         swap(first.localSize, second.localSize);
+         swap(first.storageSize, second.storageSize);
+         swap(first.localStart, second.localStart);
+         swap(first.coupling, second.coupling);
+         swap(first.neighbourSendType, second.neighbourSendType);
+         swap(first.neighbourReceiveType, second.neighbourReceiveType);
+         swap(first.data, second.data);
+      }
+
+      // Copy constructor
+      FsGrid(const FsGrid& other) : 
+         DX {other.DX},
+         DY {other.DY},
+         DZ {other.DZ},
+         physicalGlobalStart {other.physicalGlobalStart},
+         comm1d {MPI_COMM_NULL},
+         comm1d_aux {MPI_COMM_NULL},
+         comm3d {MPI_COMM_NULL},
+         comm3d_aux {MPI_COMM_NULL},
+         neighbourSendType {},
+         neighbourReceiveType {},
+         rank {other.rank}, 
+         requests {}, 
+         numRequests {0}, 
+         neighbour {other.neighbour},
+         neighbour_index {other.neighbour_index},
+         ntasksPerDim {other.ntasksPerDim},
+         taskPosition {other.taskPosition},
+         periodic {other.periodic},
+         globalSize {other.globalSize},
+         localSize {other.localSize},
+         storageSize {other.storageSize},
+         localStart {other.localStart},
+         coupling {other.coupling},
+         data {other.data}
+      {
+         if (other.comm3d != MPI_COMM_NULL) {
+            MPI_Comm_dup(other.comm3d, &comm3d);
+         }
+
+         neighbourSendType.fill(MPI_DATATYPE_NULL);
+         neighbourReceiveType.fill(MPI_DATATYPE_NULL);
+         for (int i = 0; i < neighbourSendType.size(); ++i) {
+            if (other.neighbourSendType[i] != MPI_DATATYPE_NULL) {
+               MPI_Type_dup(other.neighbourSendType[i], neighbourSendType.data() + i);
+            }
+            if (other.neighbourReceiveType[i] != MPI_DATATYPE_NULL) {
+               MPI_Type_dup(other.neighbourReceiveType[i], neighbourReceiveType.data() + i);
+            }
+         }
+      }
+
+      // Move constructor
+      // We don't have a default constructor, so just set the MPI stuff NULL
+      FsGrid(FsGrid&& other) noexcept : 
+         comm1d {MPI_COMM_NULL},
+         comm1d_aux {MPI_COMM_NULL},
+         comm3d {MPI_COMM_NULL},
+         comm3d_aux {MPI_COMM_NULL},
+         neighbourSendType {},
+         neighbourReceiveType {}
+      {
+         // NULL all the MPI stuff so they won't get freed if destroyed
+         neighbourSendType.fill(MPI_DATATYPE_NULL);
+         neighbourReceiveType.fill(MPI_DATATYPE_NULL);
+
+         using std::swap;
+         swap(*this, other);
+      }
+
+      // Copy assignment
+      FsGrid& operator=(FsGrid other) = delete;
```

Comment:
It's deleted :) 
- Canonical copy assignment is copy construct to temp and swap with it, which is probably horrible memory-wise here
- Copy-assigning by hand doesn't have the issue, but creates more bits to maintain when e.g. new variables are introduced
- We never copy assign currently. Are there any conceivable reasons to do that? Note that we do have copy _constructor_

Reply:
So you're preventing the copy assignment operator from being generated, which means anyone trying to use it won't even be able to compile? I see this does kinda conform with the rule of five, so I won't complain about that.

1) Please put the bullet points in the code - I think my question was quite valid, and there's no reason to be so obfuscated.
2) If we never copy assign in code now, then this is not an issue - right?
3) Warning: If we want to introduce e.g. unit or conformance testing in FSgrid, this will prevent us using copy assigns there either. I don't think this is a good direction to go in. 

How much work would it really be to build a manual copy assign, if you're so worried about performance?

Reply:
1. Good point.
2. Exactly.
3. I think we can cross that bridge when we get there. _Not_ having a copy constructor gives us the benefit of the code not compiling if someone tries to do something they probably don't want to (you should really think twice before copying the entire grid!).

Implementing the manual copy assign wouldn't actually be too much work. The issue is that it would essentially duplicate the copy constructor. Looking at it again it could possibly be DRY'd though so maybe it's not too bad?